### PR TITLE
[New Azure Integration Pkgs] Add DataProtection key encryptor implementation based on latest Azure.Security.KeyVault.Keys

### PR DESCRIPTION
**Pull request by pakrym**
cc @anurse @BrennanConroy

### CODE REVIEW

**Comment by pakrym** [comment]
I'm having trouble building the repo on the clean machine see https://github.com/aspnet/AspNetCore/issues/16664#issuecomment-553093953 but I was able to run tests on the project.

File: src/DataProtection/KeyVault.Keys/src/AzureDataProtectionBuilderExtensions.cs [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+            {
+                throw new ArgumentNullException(nameof(client));
+            }
+            if (string.IsNullOrEmpty(keyIdentifier))
+            {
+                throw new ArgumentException(nameof(keyIdentifier));
+            }
+
+            builder.Services.AddSingleton<IKeyEncryptionKeyResolver>(client);
+            builder.Services.Configure<KeyManagementOptions>(options =>
```
Comment (BrennanConroy) [inline src/DataProtection/KeyVault.Keys/src/AzureDataProtectionBuilderExtensions.cs]: This isn't used anywhere?

File: src/DataProtection/KeyVault.Keys/src/Properties/AssemblyInfo.cs [hunk 2]
```diff
@@ -0,0 +1,9 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System.Reflection;
+using System.Resources;
+using System.Runtime.CompilerServices;
+
+[assembly: InternalsVisibleTo("Microsoft.AspNetCore.DataProtection.KeyVault.Keys.Tests, PublicKey=0024000004800000940000000602000000240000525341310004000001000100f33a29044fa9d740c9b3213a93e57c84b472c84e0b8a0e1ae48e67a9f8f6de9d5f7f3d52ac23e48ac51801f1dc950abe901da34d2a9e3baadb141a17c77ef3c565dd5ee5054b91cf63bb3c6ab83f72ab3aafe93d0fc3c2348b764fafb0b1c0733de51459aeab46580384bf9d74c4e28164b7cde247f891ba07891c9d872ad2bb")]
```
Comment (BrennanConroy) [inline src/DataProtection/KeyVault.Keys/src/Properties/AssemblyInfo.cs]: nit: can put in csproj

File: src/DataProtection/KeyVault.Keys/test/AzureKeyVaultXmlEncryptorTests.cs [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+    public class AzureKeyVaultXmlEncryptorTests
+    {
+        [Fact]
+        public void UsesKeyVaultToEncryptKey()
+        {
+            var keyMock = new Mock<IKeyEncryptionKey>(MockBehavior.Strict);
+            keyMock.Setup(client => client.WrapKeyAsync("RSA-OAEP", It.IsAny<ReadOnlyMemory<byte>>(), default))
+                .ReturnsAsync((string _, ReadOnlyMemory<byte> data, CancellationToken __) => data.ToArray().Reverse().ToArray())
+                .Verifiable();
+
```
Comment (BrennanConroy) [inline src/DataProtection/KeyVault.Keys/test/AzureKeyVaultXmlEncryptorTests.cs]:
```suggestion
```

File: src/DataProtection/KeyVault.Keys/test/Microsoft.AspNetCore.DataProtection.KeyVault.Keys.Tests.csproj [hunk 4]
```diff
@@ -0,0 +1,14 @@
+<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <TargetFramework>$(DefaultNetCoreTargetFramework)</TargetFramework>
+    <AllowUnsafeBlocks>true</AllowUnsafeBlocks>
```
Comment (BrennanConroy) [inline src/DataProtection/KeyVault.Keys/test/Microsoft.AspNetCore.DataProtection.KeyVault.Keys.Tests.csproj]:
```suggestion
```

File: src/DataProtection/KeyVault.Keys/src/AzureDataProtectionBuilderExtensions.cs (hunk 1 above)
Comment (pakrym) [reply src/DataProtection/KeyVault.Keys/src/AzureDataProtectionBuilderExtensions.cs]: Is used by DataProtection itself https://github.com/aspnet/AspNetCore/blob/416d8cd1ea098eb643b4258247af3448f5545c32/src/DataProtection/DataProtection/src/KeyManagement/DefaultKeyResolver.cs#L48

**Comment by pranavkm** [comment]
@pakrym can you rebase on master and try: https://github.com/aspnet/AspNetCore/pull/17633?

**Review by heaths — APPROVED** [review]

File: src/DataProtection/Azure.KeyVault.Keys/src/Microsoft.AspNetCore.DataProtection.Azure.KeyVault.Keys.csproj [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+    <Description>Microsoft Azure KeyVault key encryption support.</Description>
+    <TargetFramework>netstandard2.0</TargetFramework>
+    <GenerateDocumentationFile>true</GenerateDocumentationFile>
+    <PackageTags>aspnetcore;dataprotection;azure;keyvault</PackageTags>
+    <IsShippingPackage>true</IsShippingPackage>
+    <HasReferenceAssembly>false</HasReferenceAssembly>
+  </PropertyGroup>
+
+  <ItemGroup>
+    <InternalsVisibleTo Include="Microsoft.AspNetCore.DataProtection.Azure.KeyVault.Keys.Tests" />
```
Comment (dougbu) [inline src/DataProtection/Azure.KeyVault.Keys/src/Microsoft.AspNetCore.DataProtection.Azure.KeyVault.Keys.csproj]: Suggest using one mechanism for ITVs in this project. If you choose to stick w/ this element, it takes a `Key` attribute for use in the `DynamicProxyGenAssembly2` case.

File: src/DataProtection/Azure.KeyVault.Keys/Directory.Build.props [hunk 6]
```diff
@@ -0,0 +1,10 @@
+<Project>
+  <Import Project="$([MSBuild]::GetDirectoryNameOfFileAbove($(MSBuildThisFileDirectory)..\, Directory.Build.props))\Directory.Build.props" />
+
+  <PropertyGroup>
+    <!-- These projects depend on a 3rd party source. -->
+    <ExcludeFromSourceBuild>true</ExcludeFromSourceBuild>
```
Comment (dougbu) [inline src/DataProtection/Azure.KeyVault.Keys/Directory.Build.props]: This setting should do nothing. Only ref/ projects and those in the Shared FX are built in source build.

File: src/DataProtection/Azure.KeyVault.Keys/src/AzureDataProtectionBuilderExtensions.cs [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+using Microsoft.AspNetCore.DataProtection.KeyManagement;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Identity.Client;
+
+namespace Microsoft.AspNetCore.DataProtection
+{
+    /// <summary>
+    /// Contains Azure KeyVault-specific extension methods for modifying a <see cref="IDataProtectionBuilder"/>.
+    /// </summary>
+    public static class AzureDataProtectionBuilderExtensions
```
Comment (pranavkm) [inline src/DataProtection/Azure.KeyVault.Keys/src/AzureDataProtectionBuilderExtensions.cs]:
```diff
public static class AzureDataProtectionKeyVaultKeyBuilderExtensions
```

**Comment by anurse** [comment]
Approved for March (since this is a bundle along with https://github.com/dotnet/extensions/pull/2618#issuecomment-575293740)

File: src/DataProtection/Azure.KeyVault.Keys/src/Microsoft.AspNetCore.DataProtection.Azure.KeyVault.Keys.csproj [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+  <PropertyGroup>
+    <Description>Microsoft Azure KeyVault key encryption support.</Description>
+    <TargetFramework>netstandard2.0</TargetFramework>
+    <GenerateDocumentationFile>true</GenerateDocumentationFile>
+    <PackageTags>aspnetcore;dataprotection;azure;keyvault</PackageTags>
+    <IsShippingPackage>true</IsShippingPackage>
+    <HasReferenceAssembly>false</HasReferenceAssembly>
+    <PreReleaseVersionLabel>$(AzureSDKPreReleaseVersionLabel)</PreReleaseVersionLabel>
+    <DotNetFinalVersionKind />
+    <VersionPrefix>$(AspNetCoreMajorVersion).$(AspNetCoreMinorVersion).0</VersionPrefix>
```
Comment (BrennanConroy) [inline src/DataProtection/Azure.KeyVault.Keys/src/Microsoft.AspNetCore.DataProtection.Azure.KeyVault.Keys.csproj]:
```suggestion
```

We've decided to leave the version alone, so it'll have a patch version for the first release.

**Comment by BrennanConroy** [comment]
Ping reviewers, need someone to approve.

I could do it, but that's a bit cheaty 😆

**Review by dougbu — APPROVED** [review]
Didn't review the C# but the infrastructure and .csproj files look grand

**Review by analogrelay — APPROVED** [review]
LGTM

**Comment by anurse** [comment]
Closing as these changes are moving.

**Pull request closed without merging**